### DISCUSSION TITLE
Fire event when pressure plate is pressed

### DISCUSSION THREAD

**Discussion by muelleel** [Minestom/Minestom#1248] [Help] [answered]
Hi,

I'm currently trying to get pressure plates working.
Ive tried the following:
```
        Block block = Block.HEAVY_WEIGHTED_PRESSURE_PLATE.withHandler(new BlockHandler() {
            @Override
            public @NotNull NamespaceID getNamespaceId() {
                return NamespaceID.from("test:plate");
            }

            @Override
            public void onTouch(@NotNull Touch touch) {
                System.out.println("Stepped on a plate");
            }

            @Override
            public void onDestroy(@NotNull Destroy destroy) {
                System.out.println("Handler is working");
            }
        });
```

The onDestroy gets fired when the plate is broken, but the "onTouch" does not. 
I've expected "onTouch" gets fired when an Entity steps on the plate, as the Documentation states: "Defines custom behaviour for entities touching this block."

Minestom `ab5734334c`

Am I missing something?
Thank you!

**Comment by VegetalDev** [comment]
I believe this occurs because pressure plates do not have collision, and thus are not detected as solid blocks in `Entity#touchTick()`. So you can't use a handler for this.

I think this is a bug, or at least a missing feature. As a (hacky) workaround, you could try periodically checking the block at the player's feet and invoking its handler if necessary.

**Reply by muelleel** [reply]
Thank you for this Information.

Did no one else ever tried to use pressure plates in Minestom? I's kinda off that I'm the first one asking...

My plan is to make a mini game with these plates as "mines" (over 4k plates per instance), so an hacky workaround would most likely result in an "non vanilla" feeling - which I want to avoid.

Can I maybe add my own collision box for the plates? I think this would be the "better" way, as this would already respect things like the size of the Player hitbox (or not?)

**Reply by VegetalDev** [reply]
> Can I maybe add my own collision box for the plates? I think this would be the "better" way, as this would already respect things like the size of the Player hitbox (or not?)

This sounds like it could work, but you would probably need to do that on a fork if you aren't already. 

Alternatively, a pretty simple fix to Minestom could entail `onTouch` being invoked any time the entity intersects with the same block (cubic meter) as the pressure plate. Then, in your handler, you could check if the player's vertical distance above the plate is indeed enough to warrant being triggered, keeping it in line with vanilla.

**Reply by muelleel** [reply]
Okay, since I'm not working on a Fork I'll try to implement this like you said manually.

Can you provide an basic Example for the collision with the cubic meter itself? 
Do I check that on each PlayerMoveEvent? Is this the best way?

Also the plates are not an full block wide, so in this image they should not trigger - although my hitbox is in the same Block as the plates.

![grafik](https://user-images.githubusercontent.com/40674502/179777434-b5cb3602-97f4-4e59-b410-fde3517edcfb.png)

**Comment by muelleel** [comment, marked as answer]
I've wrote this now:
It respects the height and the edges of the plate, so for most use cases it does the job! 
```
        Block plateType = Block.HEAVY_WEIGHTED_PRESSURE_PLATE;
        double pixelOffset = 0.0625;

        eventNode().addListener(PlayerMoveEvent.class, playerMoveEvent -> {
            Pos middle = playerMoveEvent.getNewPosition();

            Pos[] corners = {
                middle.add(0.3-pixelOffset , 0, 0.3-pixelOffset),
                middle.add(0.3-pixelOffset, 0, -0.3+pixelOffset),
                middle.add(-0.3+pixelOffset, 0, 0.3-pixelOffset),
                middle.add(-0.3+pixelOffset, 0, -0.3+pixelOffset)
            };

            for(Pos coroner : corners) {
                Block pressed = getBlock(coroner);
                if(pressed.compare(plateType) && playerMoveEvent.getNewPosition().y() < middle.blockY() + pixelOffset)
                    playerMoveEvent.getPlayer().sendMessage(String.valueOf(System.currentTimeMillis()));
            }

        });
```

there is certainly an faster and more efficient way to check this - if so, let me know :)

**Reply by VegetalDev** [reply]
This seems like it's as good as it's going to get without modifying any Minestom source.

Speaking of, you should consider posting an issue regarding `touchTick` not firing for non-collidable blocks. I might be able to PR a solution for it.
